README
		  *** STX Execution Pipe README ***

The STX ExecPipe library provides a convenient C++ interface to execute child
programs connected via pipes. It is a front-end to the system calls fork(),
pipe(), select() and execv() and hides all the complexity of these low-level
functions. It allows a program to build a sequence of connected children
programs with input and output of the pipe sequence redirected to a file,
string or file descriptor. The library also allows custom asynchronous data
processing classes to be inserted into the pipe or placed at source or sink of
the sequence.

An execution pipe consists of an input stream, a number of pipe stages and an
output stream. The input and output streams can be a plain file descriptor, a
file, a std::string or a special processing class. Each pipe stage is either an
executed child program or an intermediate function class. At the junction
between each stage in the pipeline the following program's stdin is connected
to the preceding stage's stdout. The input and output streams are connected to
the start and end of the pipe line.


 Input Stream                   Pipe Stages                   Output Stream
     none    |                                                |    none
      fd     |                 exec()                         |     fd
     file    |--> stage -->      or      --> stage --> ... -->|    file
    string   |              PipeFunction                      |   string
  PipeSource |                                                |  PipeSink


All this functionality is wrapped into a flexible C++ class, which can be used
in an application to construct complex sequences of external programs similar
to shell piping. Some common operations would be calls of mkisofs or tar
coupled with gzip or gpg and possibly send the output to a remote host via ssh
or ncftpput.

--- Website / API Docs / Bugs / License ---

The current source package can be downloaded from
http://idlebox.net/2010/stx-execpipe/

The classes are extensively documented using doxygen. The compiled doxygen HTML
documentation is included in the source package or can be viewed online at
http://idlebox.net/2010/stx-execpipe/stx-execpipe-0.7.1/doxygen-html/

If bugs should become known they will be posted on the above web page together
with patches or corrected versions.

The source code is released under the GNU Lesser General Public License
v2.1 (LGPL) which can be found in the file COPYING.

--- Library Usage Tutorial ---

The following tutorial shows some simple examples on how an execution pipe can
be set up.

To use the library a program must

  #include "stx-execpipe.h"

and later link against libstx-execpipe.a or include the corresponding .o / .cc
in the project's dependencies.

To run a sequence of programs you must first initialize a new ExecPipe object.
The ExecPipe object is referenced counted so you can easily pass it around
without deep-duplicating the object.

  stx::ExecPipe ep;               // creates new pipe

  stx::ExecPipe ep_ref1 = ep;     // reference to the same pipe.

Once created the input stream source can be set using one of the four
set_input_*() functions. Note that these are mutually exclusive, you must call
at most one of the following functions!

  // you can designate an existing file as input stream
  ep.set_input_file("/path/to/file");

  // or directly assign an already opened file descriptor
  int fd = ...;
  ep.set_input_fd(fd);

  // or pass the contents of a std::string as input
  std::string str = ...;
  ep.set_input_string(&str);

  // or attach a data generating source class (details later).
  PipeSource source;
  ep.set_input_source(&source);

The input stream objects are _not_ copied. The fd, string or source object must
still exist when calling run().

After setting up the input you specify the individual stages in the pipe by
adding children programs to exec() or function classes. The stx::ExecPipe
provides different variants of add_exec*(), which are derived from the exec*()
system call variants.

  // add simple exec() call with full path.
  ep.add_exec("/bin/cat");

  // add exec() call with up to three direct parameters.
  ep.add_exec("/bin/echo", "one", "two", "three");

  // add exec() call with many parameters. the vector is _not_ copied.
  std::vector<std::string> tarargs;
  tarargs.push_back("/bin/tar");
  tarargs.push_back("--create");
  tarargs.push_back("--verbose");
  tarargs.push_back("--gzip");
  tarargs.push_back("--file");
  tarargs.push_back("/path/to/file");
  ep.add_exec(&tarargs);

  // add execp() call which searches $PATH. see man 3 execvp.
  ep.add_execp("cat");

  // same with up to three parameters.
  ep.add_execp("echo", "one", "two", "three");

  // and also works with a vector of arguments.
  ep.add_execp(&tarargs);

  // most versatile function: call execve() with program name, argv[] arguments
  // and a set of environment variables.
  std::vector<std::string> gzipargs;
  gzipargs.push_back("gunzip");           // this changes argv[0]

  std::vector<std::string> gzipenvs;      // set environment variable
  gzipenvs.push_back("GZIP=-d --name");

  ep.add_exece("/bin/gzip", &gzipargs, &gzipenvs);

  // insert an intermediate data processing class into the pipe (details
  later).
  PipeFunction function;
  ep.add_function(&function);

After configuring the pipe stages the user program can redirect the pipe's
output using one of the four set_output_*() functions. These correspond
directly the to input functions.

  // designate a file as output, it will be over-written,
  ep.set_output_file("/path/to/file");

  // or directly assign an already opened file descriptor
  int fd = ...;
  ep.set_output_fd(fd);

  // or save output in a std::string object
  std::string str = ...;
  ep.set_output_string(&str);

  // or attach a sink class (details later).
  PipeSink sink;
  ep.set_output_sink(&sink);

The three steps above can be done in any order. Once the pipeline is configured
as required, a call to run() will set up the input and output file descriptors,
launch all children programs, wait until these finish and concurrently process
data passed between parent and children.

If any system calls fail while running the pipe, the run() function will throw
() a std::runtime_error exception. So wrap run() in a try-catch block.

  try {
      ep.run();
  }
  catch (std::runtime_error &e) {
      std::cerr << "Pipe execution failed: " << e.what() << std::endl;
  }

After running all children their return status should be checked. These can be
inspected using the following functions. The integer parameter specifies the
exec stage in the pipe sequence.

  // get plain return status as indicated by wait().
  int rs = ep.get_return_status(0)

  // get return code for normally terminated program.
  int rc = ep.get_return_code(1);

  // get signal for abnormally terminated program (like segfault).
  int rg = ep.get_return_signal(1);

Most program have a return code of 0 when no error occurred. Therefore, a
convenience function is available which checks whether all program stages
returned zero. This is what would usually be used.

  // check all that program returned zero
  if (ep.all_return_codes_zero()) {
      // run was ok.
  }
  else {
      // error handling.
  }

After checking the return error codes the pipe's results can be used.
The tarball contains three simple examples of using the different exec()
variants and input/output redirections. See examples/simple1.cc, examples/
simple2.cc or examples/simple3.cc. More a more elaborate example using data
processing classes see the continued tutorial below.

--- Data Processing Classes ---

One of the big features of the STX ExecPipe classes is the ability to insert
intermediate asynchronous data processing classes into the pipe sequence. The
data of the pipe line is returned to the parent process and, after arbitrary
computations, can be sent on to the following execution stages. Besides
intermediate processing, the input and output stream can be attached to source
or sink classes.

This feature can be used to generate input data, e.g. binary data or file
listing, or peek at the data flowing between stages, e.g. to compute a SHA1
digest, or to directly processes output data while the children are running.

The data processing classes must be derived from one of the three abstract
classes: stx::PipeSource for generating input streams, stx::PipeFunction for
intermediate processing between stages or stx::PipeSink for receiving output.

For generating an input stream a class must derive from stx::PipeSource and
implement the poll() function. This function is called when new data can be
pushed into the pipe. When poll() is called, new data must be generated and
delivered via the write() function of stx::PipeSource. If more data is
available poll() must return true, otherwise the input stream is terminated.

Intermediate data processing classes must derive from stx::PipeFunction and
implement the two pure virtual function process() and eof(). As the name
suggests, data is delivered to the class via the process() function. After
processing the data it may be forwarded to the next pipe stage via the
inheritedwrite() function. Note that the library does not automatically forward
data, so if you forget to write() data, then the following stage does not
receive anything. When the preceding processing stage closes its data stream
the function eof() is called.

To receive the output stream a class must derive from stx::PipeSink. Similar to
stx::PipeFunction, an output sink must implement the two pure virtual function
process() and eof(). However, different from an intermediate class the stx::
PipeSink does not provide a write() function, so no data can be forwarded.

For a full example of using stx::PipeSource to iterate through a file list and
stx::PipeFunction to compute an intermediate SHA1 digest see examples/
functions1.cc.

-------------------------------------------------------------------------------

Author: Timo Bingmann (Mail: tb a-with-circle idlebox dot net)
Date: 2010-07-18


ChangeLog
2010-07-30	Timo Bingmann
	* configure.ac: added AC_SYS_LARGEFILE to enable in/output to
	large files.
	* stx-execpipe.cc: also added _FILE_OFFSET_BITS here for direct
	compilation and linking.

2010-07-18	Timo Bingmann
	initial release.


testsuite/test_execpipe.cc
// -*- mode: c++; fill-column: 79 -*-
// $Id$

/*
 * STX Execution Pipe Library v0.7.1
 * Copyright (C) 2010 Timo Bingmann
 *
 * This library is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as published by the
 * Free Software Foundation; either version 2.1 of the License, or (at your
 * option) any later version.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License
 * for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this library; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
 */

#include "stx-execpipe.h"

#include <assert.h>
#include <iostream>
#include <sstream>
#include <iomanip>

// Test pipe: none -> program -> string
void test_none_program_string()
{
    stx::ExecPipe ep;

    std::string output;
    ep.set_output_string(&output);

    ep.add_exec("/bin/echo", "test123");

    assert( ep.run().all_return_codes_zero() );

    assert( output == "test123\n" );
}

// Test pipe: string -> program -> string
void test_string_program_string()
{
    stx::ExecPipe ep;

    std::string input = "test123";
    input += std::string(1024*1024, '\1');
    ep.set_input_string(&input);

    std::string output;
    ep.set_output_string(&output);

    ep.add_exec("/bin/cat");

    assert( ep.run().all_return_codes_zero() );

    //std::cout << "o " << output.size() << " i " << input.size() << "\n";

    assert( output == input );
}

// Test pipe: string -> program -> program -> string
void test_string_program_program_string()
{
    stx::ExecPipe ep;

    std::string input = "test123";
    ep.set_input_string(&input);

    std::string output;
    ep.set_output_string(&output);

    ep.add_exec("/bin/cat");
    ep.add_execp("md5sum");

    assert( ep.run().all_return_codes_zero() );

    assert( output == "cc03e747a6afbbcbf8be7668acfebee5  -\n" );
}

// Test pipe: file -> program -> string
void test_file_program_string()
{
    stx::ExecPipe ep;

    ep.set_input_file("/proc/uptime");

    std::string output;
    ep.set_output_string(&output);

    ep.add_execp("sort");

    assert( ep.run().all_return_codes_zero() );

    assert( output.size() );
}

// Test pipe: string -> program -> object

class TestSink : public stx::PipeSink
{
public:
    std::string		m_save;

    bool		m_ok;

    TestSink()
	: m_ok(false)
    {
    }

    virtual void process(const void* data, unsigned int datalen)
    {
	m_save.append(reinterpret_cast<const char*>(data), datalen);
    }

    virtual void eof()
    {
	m_ok = (m_save == "cc03e747a6afbbcbf8be7668acfebee5  -\n");
    }
};

void test_string_program_object()
{
    stx::ExecPipe ep;

    std::string input = "test123";
    ep.set_input_string(&input);

    TestSink sink;
    ep.set_output_sink(&sink);

    ep.add_execp("md5sum");

    assert( ep.run().all_return_codes_zero() );

    assert( sink.m_ok );
}

class TestSource : public stx::PipeSource
{
public:
    unsigned int	m_count;

    std::string		m_wrote;

    TestSource()
	: m_count(100*1024)
    {
    }

    virtual bool poll()
    {
	for (unsigned int i = 0; i < 1000 && m_count > 0; ++i, --m_count)
	{
	    write(&i, sizeof(unsigned char));

	    m_wrote.append(reinterpret_cast<const char*>(&i), sizeof(unsigned char));
	}

	return (m_count > 0);
    }
};

// Test pipe: object -> program -> string

void test_object_program_string()
{
    stx::ExecPipe ep;

    TestSource source;
    ep.set_input_source(&source);

    std::string output;
    ep.set_output_string(&output);

    ep.add_execp("cat");

    assert( ep.run().all_return_codes_zero() );

    assert( source.m_wrote == output );
}

// Test pipe: object -> program -> function -> program -> string

#include <openssl/md5.h>

class TestFunctionMD5 : public stx::PipeFunction
{
public:

    MD5_CTX 	m_ctx;

    std::string m_digest;

    TestFunctionMD5()
    {
	MD5_Init(&m_ctx);
    }

    virtual void process(const void* data, unsigned int datalen)
    {
	MD5_Update(&m_ctx, data, datalen);

	write(data, datalen);
    }

    virtual void eof()
    {
	unsigned char digest[MD5_DIGEST_LENGTH];
	MD5_Final(digest, &m_ctx);

	m_digest.assign(reinterpret_cast<char*>(digest), sizeof(digest));
    }
};

std::string HexString(const std::string& str)
{
    std::ostringstream os;

    os << std::hex << std::setfill('0');

    for (std::string::const_iterator si = str.begin();
	 si != str.end(); ++si)
    {
	os << std::setw(2) << (static_cast<unsigned int>(*si) & 0xFF);
    }

    return os.str();
}

void test_object_program_object_program_string()
{
    stx::ExecPipe ep;

    TestSource source;
    ep.set_input_source(&source);

    std::string output;
    ep.set_output_string(&output);

    ep.add_execp("cat");

    TestFunctionMD5 func;
    ep.add_function(&func);

    ep.add_execp("md5sum");

    assert( ep.run().all_return_codes_zero() );

    assert( HexString(func.m_digest) == "0b66fcadf3a46cce7184487c4dabaf0f" );
    assert( output == "0b66fcadf3a46cce7184487c4dabaf0f  -\n" );
}

void test_object_program_object_string()
{
    stx::ExecPipe ep;

    TestSource source;
    ep.set_input_source(&source);

    std::string output;
    ep.set_output_string(&output);

    ep.add_execp("cat");

    TestFunctionMD5 func;
    ep.add_function(&func);

    assert( ep.run().all_return_codes_zero() );

    assert( HexString(func.m_digest) == "0b66fcadf3a46cce7184487c4dabaf0f" );
    assert( output.size() == 100*1024 );
}

void test_none_program_set_string()
{
    stx::ExecPipe ep;

    std::vector<std::string> args;
    args.push_back("/bin/sh");
    args.push_back("-c");
    args.push_back("set");

    std::vector<std::string> envs;
    envs.push_back("TEST=123");

    ep.add_exece("/bin/sh", &args, &envs);

    std::string output;
    ep.set_output_string(&output);

    assert( ep.run().all_return_codes_zero() );

    assert( output.find("TEST=123") != std::string::npos );
}

void test_error_debug_output_null(const char*)
{
}

void test_error_none_program_none()
{
    stx::ExecPipe ep;
    ep.set_debug_level(stx::ExecPipe::DL_INFO);
    ep.set_debug_output(test_error_debug_output_null);

    ep.add_exec("xyz-non-existing-program");

    ep.run();

    assert( !ep.all_return_codes_zero() );

    assert( ep.get_return_code(0) == 255 );
}

void test_segfault_none_program_none()
{
    stx::ExecPipe ep;

    ep.add_exec("./test_segfault");

    ep.run();
    assert( !ep.all_return_codes_zero() );

    assert( ep.get_return_signal(0) == 11 );
}

int main()
{
    test_none_program_string();
    test_string_program_string();
    test_string_program_program_string();
    test_file_program_string();
    test_string_program_object();
    test_object_program_string();
    test_object_program_object_program_string();
    test_object_program_object_string();
    test_none_program_set_string();

    test_error_none_program_none();
    test_segfault_none_program_none();

    return 0;
}


testsuite/test_segfault.cc
// -*- mode: c++; fill-column: 79 -*-
// $Id$

/*
 * STX Execution Pipe Library v0.7.1
 * Copyright (C) 2010 Timo Bingmann
 *
 * This library is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as published by the
 * Free Software Foundation; either version 2.1 of the License, or (at your
 * option) any later version.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License
 * for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this library; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
 */

/* This program intensionally segfaults as a test for test_execpipe. */

int main()
{
    char* nullptr = 0;

    *nullptr = 1;

    return 0;
}


testsuite/test_ringbuffer.cc
// -*- mode: c++; fill-column: 79 -*-
// $Id$

/*
 * STX Execution Pipe Library v0.7.1
 * Copyright (C) 2010 Timo Bingmann
 *
 * This library is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as published by the
 * Free Software Foundation; either version 2.1 of the License, or (at your
 * option) any later version.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License
 * for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this library; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
 */

/*
 * Extensive test case for internal RingBuffer. Covers both internal
 * states.
 */

#include "stx-execpipe.h"

// need access to anonymous namespace
#include "stx-execpipe.cc"

void test1()
{
    stx::RingBuffer rb;

    for (unsigned int i = 0; i < 128; ++i)
    {
	rb.write(&i, sizeof(i));
    }

    assert( rb.size() == 128 * sizeof(unsigned int) );
    
    for (unsigned int i = 0; i < 128; ++i)
    {
	assert( rb.bottomsize() >= sizeof(unsigned int) );
	assert( *reinterpret_cast<unsigned int*>(rb.bottom()) == i );

	rb.advance( sizeof(unsigned int) );
    }

    assert( rb.size() == 0 );
    assert( rb.bottomsize() == 0 );
    assert( rb.buffsize() == 1024 );

    for (unsigned int i = 0; i < 512; ++i)
    {
	assert( rb.size() == 0 );
	assert( rb.bottomsize() == 0 );

	rb.write(&i, sizeof(i));

	assert( rb.bottomsize() == 4 );
	assert( *reinterpret_cast<unsigned int*>(rb.bottom()) == i );

	rb.advance( sizeof(unsigned int) );
    }
}

void test2()
{
    char buffer[2048];

    for (unsigned int i = 0; i < sizeof(buffer); ++i)
	buffer[i] = i;

    // test growth in first buffer state
    {
	stx::RingBuffer rb;

	rb.write(buffer, 256);
	rb.advance(256);

	assert(rb.size() == 0);
	assert(rb.bottomsize() == 0);
	assert(rb.buffsize() == 1024);

	rb.write(buffer, 512);

	assert(rb.size() == 512);
	assert(rb.bottomsize() == 512);
	assert(rb.buffsize() == 1024);

	rb.write(buffer, 1024);

	assert(rb.size() == 512+1024);
	assert(rb.bottomsize() == 512+1024);
	assert(rb.buffsize() == 2048);
    }

    // test growth in second buffer state
    {
	stx::RingBuffer rb;

	rb.write(buffer, 512+256);
	rb.advance(512+256);

	assert(rb.size() == 0);
	assert(rb.bottomsize() == 0);
	assert(rb.buffsize() == 1024);

	rb.write(buffer, 512);

	assert(rb.size() == 512);
	assert(rb.bottomsize() == 256);
	assert(rb.buffsize() == 1024);

	rb.write(buffer, 1024);

	assert(rb.size() == 512+1024);
	assert(rb.bottomsize() == 256);
	assert(rb.buffsize() == 2048);

    }
}

int main()
{
    test1();
    test2();

    return 0;
}


examples/simple1.cc
// -*- mode: c++; fill-column: 79 -*-
// $Id$

/*
 * STX Execution Pipe Library v0.7.1
 * Copyright (C) 2010 Timo Bingmann
 *
 * This library is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as published by the
 * Free Software Foundation; either version 2.1 of the License, or (at your
 * option) any later version.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License
 * for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this library; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
 */

/*
 * This first example shows how to directly call a program, in this case
 * "/bin/echo" and save it's output in a std::string.
 */

#include "stx-execpipe.h"

#include <iostream>
#include <stdexcept>

int main()
{
    stx::ExecPipe ep;

    ep.add_exec("/bin/echo", "-n", "test123");

    std::string output;
    ep.set_output_string(&output);

    try {
	if (!ep.run().all_return_codes_zero())
	{
	    std::cout << "Error calling echo: return code = " << ep.get_return_code(0) << std::endl;
        }
	else
	{
	    std::cout << "echo output: " << output << std::endl;
	}
    }
    catch (std::runtime_error &e)
    {
	std::cout << "Error running children: " << e.what() << std::endl;
    }

    return 0;
}



examples/simple2.cc
// -*- mode: c++; fill-column: 79 -*-
// $Id$

/*
 * STX Execution Pipe Library v0.7.1
 * Copyright (C) 2010 Timo Bingmann
 *
 * This library is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as published by the
 * Free Software Foundation; either version 2.1 of the License, or (at your
 * option) any later version.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License
 * for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this library; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
 */

/*
 * This second example shows how to directly call a program, but this time use
 * execp() variants, which search the PATH environment.
 */

#include "stx-execpipe.h"

#include <iostream>
#include <stdexcept>

int main()
{
    stx::ExecPipe ep;

    std::string input = "test123";
    ep.set_input_string(&input);

    ep.add_execp("sha1sum");

    std::string output;
    ep.set_output_string(&output);

    try {
	if (!ep.run().all_return_codes_zero())
	{
	    std::cout << "Error calling echo: return code = " << ep.get_return_code(0) << std::endl;
        }
	else
	{
	    std::cout << "sha1sum output: " << output << std::endl;
	}
    }
    catch (std::runtime_error &e)
    {
	std::cout << "Error running children: " << e.what() << std::endl;
    }

    return 0;
}


examples/simple3.cc
// -*- mode: c++; fill-column: 79 -*-
// $Id$

/*
 * STX Execution Pipe Library v0.7.1
 * Copyright (C) 2010 Timo Bingmann
 *
 * This library is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as published by the
 * Free Software Foundation; either version 2.1 of the License, or (at your
 * option) any later version.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License
 * for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this library; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
 */

/*
 * This third example shows how to call a sequence of programs. The pipe
 * consists of "ls --size /bin" listing a direction, grepping out all 'shells'
 * and sorting these by file size.
 */

#include "stx-execpipe.h"

#include <iostream>
#include <stdexcept>

int main()
{
    stx::ExecPipe ep;

    ep.add_execp("ls", "--size", "/bin");

    ep.add_execp("grep", "sh");

    std::vector<std::string> sortargs;
    sortargs.push_back("sort");
    sortargs.push_back("--numeric-sort");
    sortargs.push_back("--ignore-leading-blanks");
    sortargs.push_back("--reverse");
    sortargs.push_back("--stable");
    ep.add_execp(&sortargs);

    std::string output;
    ep.set_output_string(&output);

    try {
	if (!ep.run().all_return_codes_zero())
	{
	    std::cout << "Error calling programs: " << std::endl
		      << "ls returned = " << ep.get_return_code(0) << std::endl
		      << "grep returned = " << ep.get_return_code(1) << std::endl
		      << "sort returned = " << ep.get_return_code(2) << std::endl;
        }
	else
	{
	    std::cout << "pipe output: " << output << std::endl;
	}
    }
    catch (std::runtime_error &e)
    {
	std::cout << "Error running children: " << e.what() << std::endl;
    }

    return 0;
}


examples/functions1.cc
// -*- mode: c++; fill-column: 79 -*-
// $Id$

/*
 * STX Execution Pipe Library v0.7.1
 * Copyright (C) 2010 Timo Bingmann
 *
 * This library is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as published by the
 * Free Software Foundation; either version 2.1 of the License, or (at your
 * option) any later version.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License
 * for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this library; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
 */

/*
 * This example shows how to use the function classes stx::PipeSource and
 * stx::PipeFunction to insert custom processing into a pipe sequence. The
 * application calls tar to create an archive, calculates the SHA1 digest of
 * the uncompressed tarball and then pipes the data into gzip for compression.
 */

#include "stx-execpipe.h"

#include <iostream>
#include <sstream>
#include <iomanip>
#include <stdexcept>

#include <openssl/sha.h>

class FilelistSource : public stx::PipeSource
{
public:
    
    // List of files to send to tar.
    std::vector<std::string>	m_list;

    // Current position in list.
    unsigned int	m_pos;
    
    FilelistSource()
	: m_pos(0)
    {
    }

    // Send one file name each time polled.
    virtual bool poll()
    {
	if (m_pos < m_list.size())
	{
	    write(m_list[m_pos].data(), m_list[m_pos].size());
	    write("\n", 1);
	    ++m_pos;
	}

	return (m_pos < m_list.size());
    }
};

class Sha1Function : public stx::PipeFunction
{
public:

    // Context of running SHA1 digest
    SHA_CTX 	m_ctx;

    // Finished digest generated in eof().
    std::string m_digest;

    Sha1Function()
    {
	SHA1_Init(&m_ctx);
    }

    // Update the sha1 digest context and pass on unmodified data.
    virtual void process(const void* data, unsigned int datalen)
    {
	SHA1_Update(&m_ctx, data, datalen);

	write(data, datalen);
    }

    // Calculate final SHA1 digest once the data stream closes.
    virtual void eof()
    {
	unsigned char digest[SHA_DIGEST_LENGTH];
	SHA1_Final(digest, &m_ctx);

	m_digest.assign(reinterpret_cast<char*>(digest), sizeof(digest));
    }
};

// Converts a binary string into hexadecimal digits.
std::string HexString(const std::string& str)
{
    std::ostringstream os;
    os << std::hex << std::setfill('0');

    for (std::string::const_iterator si = str.begin(); si != str.end(); ++si)
    {
	os << std::setw(2) << (static_cast<unsigned int>(*si) & 0xFF);
    }

    return os.str();
}

int main()
{
    stx::ExecPipe ep;

    // initialize a source object generating some file names. obviously in a
    // real application this list would be longer.
    FilelistSource source;

    source.m_list.push_back("/bin/sh");
    source.m_list.push_back("/bin/bash");
    source.m_list.push_back("/bin/ls");
    source.m_list.push_back("/bin/gzip");

    ep.set_input_source(&source);

    // add new exec stage calling tar with an option to read files from stdin
    std::vector<std::string> tarargs;

    tarargs.push_back("tar");
    tarargs.push_back("--create");
    tarargs.push_back("--verbose");
    tarargs.push_back("--no-recursion");
    tarargs.push_back("--files-from");
    tarargs.push_back("/dev/stdin");

    ep.add_execp(&tarargs);

    // insert an intermediate processing stage to save the SHA1 sum of the
    // uncompressed tarball.
    Sha1Function sha1tar;

    ep.add_function(&sha1tar);

    // add compression stage
    ep.add_execp("gzip", "-9");

    // set output stream to a temporary file
    ep.set_output_file("/tmp/stx-execpipe-functions1.tar.gz");

    // run pipe
    try {
	if (!ep.run().all_return_codes_zero())
	{
	    std::cout << "Error calling programs: " << std::endl
		      << "tar returned = " << ep.get_return_code(0) << std::endl
		      << "gzip returned = " << ep.get_return_code(2) << std::endl;
        }
	else
	{
	    std::cout << "SHA1 of uncompress tar: " << HexString(sha1tar.m_digest) << std::endl
		      << "You can verify the digest using:" << std::endl
		      << "    zcat /tmp/stx-execpipe-functions1.tar.gz | sha1sum" << std::endl;
	}
    }
    catch (std::runtime_error &e)
    {
	std::cout << "Error running children: " << e.what() << std::endl;
    }

    return 0;
}
